pandas/core/missing.py

```diff
@@ -620,144 +618,65 @@ def interpolate_2d(
     ----------
     values: array-like
         Input array.
-    method: str, default "pad"
-        Interpolation method. Could be "bfill" or "pad"
+    method: {"pad", "backfill", "ffill", "bfill"}, default "pad"
+        Interpolation method.
     axis: 0 or 1
         Interpolation axis
     limit: int, optional
         Index limit on interpolation.
     limit_area: str, optional
-        Limit area for interpolation. Can be "inside" or "outside"
+        Limit area for interpolation. Can be "inside" or "outside".
+    mask: numpy array, optional
+        boolean array of values to be interpolated.
 
     Returns
     -------
-    values: array-like
-        Interpolated array.
+    tuple of values: array-like, mask: array or None
+        Interpolated array and updated mask.
     """
     if limit_area is not None:
-        return np.apply_along_axis(
-            partial(
-                _interpolate_with_limit_area,
-                method=method,
-                limit=limit,
-                limit_area=limit_area,
+        return (
+            np.apply_along_axis(
+                partial(
+                    _interpolate_with_limit_area,
+                    method=method,
+                    limit=limit,
+                    limit_area=limit_area,
+                ),
+                axis,
+                values,
             ),
-            axis,
-            values,
+            mask,
         )
 
-    orig_values = values
-
-    transf = (lambda x: x) if axis == 0 else (lambda x: x.T)
+    if not np.all(values.shape):
+        return values, mask
 
-    # reshape a 1 dim if needed
-    ndim = values.ndim
-    if values.ndim == 1:
-        if axis != 0:  # pragma: no cover
-            raise AssertionError("cannot interpolate on a ndim == 1 with axis != 0")
-        values = values.reshape(tuple((1,) + values.shape))
-
-    method = clean_fill_method(method)
-    tvalues = transf(values)
-    if method == "pad":
-        result, _ = _pad_2d(tvalues, limit=limit)
-    else:
-        result, _ = _backfill_2d(tvalues, limit=limit)
-
-    result = transf(result)
-    # reshape back
-    if ndim == 1:
-        result = result[0]
-
-    if orig_values.dtype.kind in ["m", "M"]:
-        # convert float back to datetime64/timedelta64
-        result = result.view(orig_values.dtype)
-
-    return result
-
-
-def _fillna_prep(values, mask=None):
-    # boilerplate for _pad_1d, _backfill_1d, _pad_2d, _backfill_2d
+    orig_values, orig_mask = values, mask
 
     if mask is None:
         mask = isna(values)
 
-    mask = mask.view(np.uint8)
-    return mask
+    if axis == 1:
+        values, mask = values.T, mask.T
 
+    # reshape a 1 dim if needed
+    if values.ndim == 1:
+        if axis != 0:
+            raise AssertionError("cannot interpolate on a ndim == 1 with axis != 0")
+        values, mask = values[np.newaxis, :], mask[np.newaxis, :]
 
-def _datetimelike_compat(func: F) -> F:
```

Comment:
can we keep datetimelike_compat and apply it to interpolate_2d (i learned in #40294 that ATM interpolate_2d doesnt need it ATM bc it the wrapper is used in the functions that this PR removes)

Reply:
can do. although as stated recently from my experience with experimenting with numba, these wrappers would probably be best in _libs adjacent to the cython code to give a cleaner api and remove conversions needed for the cython code from the python code

Reply:
im looking forward to seeing this, increasingly positive on the idea of doing more numba and less cython

Reply:
slow down. not discussed this yet. My primary motivations are no_arch releases, fully typed libraries without stubs and utilization of more cpu cores and maybe gpu with standard pandas (AFAIK cudf/rapids are planning to rely on wsl2 in order to support Windows and not support Windows natively).

I've only just started, so don't know if this is a realistic goal or whether a more incremental approach of using numba when installed for parallelization is more feasible.

Reply:
i reserve the right to be enthused about this idea

Reply:
sure. i'm enthusiastic too at this early stage...

- Inline typing compatible with mypy
- Faster local development cycle
- Simpler development environment for new contributors
- No wheel building, maybe faster release cycles
- Maybe can add some basic gpu support in future
- Benefits of numba
- Less complex code 
- No need for some prep functions, casting, asi8, ensure_int/float etc
- Style checks apply to library code.
- Let numba handle types and type conversions
- Let numba handle memory allocation and deallocation

for the basic functions i've looked at so far, the performance of numba seems to be comparable to cython when comparing to a pure python implementation (but i've not touched anything yet where we have C classes)

Reply:
oh numba perforance can often be much better than cython, and usually not less than. The issue we have is that we cannot move much of the core functions as we do not have a strict dependency at this point. we can certainly revisit things going forward.